### PR TITLE
Initial Dockerization attempt

### PR DESCRIPTION
**Pull request by zovt**
This uses a multi-stage build: one for building `onebusaway-gtfs-modules` and OpenTripPlanner, and another which is the actual image we use to run it.

To build both of these, we're using Java binaries provided by the Adoptium project (https://adoptium.net/), which is the rebrand of the AdoptOpenJDK project. We currently have the AdoptOpenJDK versions listed in the `.tool-versions` file, but they actually are no longer available.

Also of note: `onebusaway-gtfs-modules` builds properly with JDK17, but unfortunately our version of OpenTripPlanner does not. It does build with JDK8, so we fetch both versions.

We also pull a slightly older version of Maven, since the latest version of Maven defaults to disallowing `http://` repositories. There may be some quick fixes we can do to the `pom.xml` file which will allow us to use the latest version.

The actual build steps are mostly copied verbatim from this repo's `README`, minus the stuff about `asdf` (since as noted before, the versions we specify there no longer exist). First, we clone and build `onebusaway-gtfs-modules`, and then we copy the files from the local `OpenTripPlanner` directory into the container and build that.

Once `OpenTripPlanner` is built, we begin a new stage that will become the image we ultimately deploy. We fetch the Adoptium JRE8 (not JDK), and then copy the necessary files from the build container (NOTE: the files we copy are the same ones listed in `make_deploy.sh`).

The resulting image has a default `ENTRYPOINT` that will run `OpenTripPlanner` (using the same command as in `Procfile`). The image exposes port `5000`, which can be mapped to the appropriate port on the host, serving plain HTTP traffic.

On my machine, the resulting image is 771MB. Around ~80MB comes from the `debian:stable-slim` image that we're using for the base, around ~100MB comes from the JRE, and the rest is the `OpenTripPlanner` application + data. It may be possible to find a slimmer image for the base, I just went with Debian because that's a distribution that I'm familiar with and is typically fairly stable.

As far as testing this goes, I was able to point my local `dotcom` instance at the port on my local machine, and a couple of trips I checked using the Trip Planner functionality on the website seemed to match what I see on prod, however I'm not super familiar with what is using OTP so it's worth doing some more testing.

Open questions:
- Do we want to expose the plain HTTP port from the Docker image to the host, or should we expose the HTTPS port?
- Is there a slimmer / better base image for the final image?

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+RUN curl -Lo jdk8.tar.gz https://github.com/adoptium/temurin8-binaries/releases/download/jdk8u322-b06/OpenJDK8U-jdk_x64_linux_hotspot_8u322b06.tar.gz
+RUN tar xvf jdk8.tar.gz && rm jdk8.tar.gz
+ENV JAVA_HOME_JDK_8=/java/jdk8u322-b06/
+
+RUN curl -Lo maven.tar.gz https://dlcdn.apache.org/maven/maven-3/3.6.3/binaries/apache-maven-3.6.3-bin.tar.gz
+RUN tar xvf maven.tar.gz && rm maven.tar.gz
+ENV PATH="/java/apache-maven-3.6.3/bin/:$PATH"
+
+WORKDIR /build/onebusaway-gtfs-modules
+RUN git clone https://github.com/mbta/onebusaway-gtfs-modules .
```
Comment (paulswartz) [inline Dockerfile:19]: nitpick: you don't need the full git history here, so you can use `--depth 1` to only get the latest data. 
```suggestion
RUN git clone --depth 1 https://github.com/mbta/onebusaway-gtfs-modules .
```

File: Dockerfile [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+WORKDIR /build/OpenTripPlanner
+ADD . .
+
+RUN ./mbta/update_gtfs.sh
+RUN ./mbta/update_pbf.sh
+ENV JAVA_HOME="$JAVA_HOME_JDK_8" PATH="$JAVA_HOME_JDK_8/bin:$PATH"
+RUN ./mbta/build.sh
+
+FROM debian:stable-slim as runner
+RUN apt-get update && apt-get upgrade -y && apt-get install -y --no-install-recommends curl ca-certificates
```
Comment (paulswartz) [inline Dockerfile:31]: suggestion: if you also fetch the JRE in the builder, you can `COPY` it over into the runner without needing to install `curl` and `ca-certificates`.

**Review by paulswartz — COMMENTED** [review]
Thank you for this! I was also able to build this on my local machine.

There are some additional `Dockerfile` best practices in the New Application checklist: https://www.notion.so/mbta-downtown-crossing/Early-Development-be06b5a2c1404056a32c9dda64a09c04#76164e73fab5459ab528b82ecf58523b

**Comment by paulswartz** [comment]
Some additional best practices:
- [ ] create an unprivileged user to run the application
- [ ] run under `dumb-init`

**Comment by ianwestcott** [comment]
> Do we want to expose the plain HTTP port from the Docker image to the host, or should we expose the HTTPS port?

HTTP is fine, we usually handle HTTPS termination at the load balancer.

> Is there a slimmer / better base image for the final image?

You could try an Alpine image maybe? But I'm not sure how much of a difference it would make. I'm unable to build the container locally (due to memory issues with Colima I think) but inspecting the Beanstalk artifacts I see that the `/var/graphs/mbta/Graph.obj` file makes up the bulk of the size of the built image (~460mb). Given that these are always going to be big images, I'm not terribly concerned with splitting hairs over the size of the base image.

Anyway this looks great to me, thanks for taking it on! I'll let @paulswartz give final approval. The next step will be to see if you can get this to build in GitHub Actions; hopefully you won't encounter anything like the resource constraints I seem to have on my workstation. 🤞

**Comment by paulswartz** [comment]
@ianwestcott I needed to create a new Colima VM with 8GB of memory:
```
$ colima start 8gb -c 2 -d 20 -m 8
$ docker build .
```

**Review by paulswartz — APPROVED** [review]
🍰

**Pull request merged**